core/modules/catalog/store/product/actions.ts

```diff
@@ -585,76 +585,58 @@ const actions: ActionTree<ProductState, RootState> = {
-        if (isServer) {
-          subloaders.push(context.dispatch('setupBreadcrumbs', { product: product }))
-          subloaders.push(context.dispatch('filterUnavailableVariants', { product: product }))
-        } else {
-          attributesPromise.then(() => context.dispatch('setupBreadcrumbs', { product: product })) // if this is client's side request postpone breadcrumbs setup till attributes are loaded to avoid too-early breadcrumb switch #2469
-          context.dispatch('filterUnavailableVariants', { product: product }) // exec async
-        }
-        subloaders.push(attributesPromise)
-
-        // subloaders.push(context.dispatch('setupVariants', { product: product })) -- moved to "product/single"
-        /* if (product.type_id === 'grouped' || product.type_id === 'bundle') { -- moved to "product/single"
-          subloaders.push(context.dispatch('setupAssociated', { product: product }).then((subloaderresults) => {
-            context.dispatch('setCurrent', product) // because setup Associated can modify the product price we need to update the current product
-          }))
-        } */
-
-        context.dispatch('setProductGallery', { product: product })
-
-        if (config.products.preventConfigurableChildrenDirectAccess) {
-          subloaders.push(context.dispatch('checkConfigurableParent', { product: product }))
-        }
-      } else { // error or redirect
+    const product = await dispatch('single', { options: productSingleOptions })
+    if (product.status >= 2) {
+      throw new Error(`Product query returned empty result product status = ${product.status}`)
+    }
+    if (product.visibility === 1) { // not visible individually (https://magento.stackexchange.com/questions/171584/magento-2-table-name-for-product-visibility)
+      throw new Error(`Product query returned empty result product visibility = ${product.visibility}`)
+    }
 
-      }
-      return subloaders
+    const productFields = Object.keys(product).filter(fieldName => {
+      return config.entities.product.standardSystemFields.indexOf(fieldName) < 0 // don't load metadata info for standard fields
     })
+    await dispatch('attribute/list', { // load attributes to be shown on the product details - the request is now async
+      filterValues: config.entities.product.useDynamicAttributeLoader ? productFields : null,
+      only_visible: config.entities.product.useDynamicAttributeLoader === true,
+      only_user_defined: true,
+      includeFields: config.entities.optimize ? config.entities.attribute.includeFields : null
+    }, { root: true })
+    await Promise.all([
+      dispatch('setupBreadcrumbs', { product: product }),
+      dispatch('filterUnavailableVariants', { product: product }),
+      dispatch('setProductGallery', { product: product })
+    ])
```

Comment:
By putting await on attributePromise and then adding `setProductGallery`  and `filterUnavailableAttributes` to Promise.all we're slowering down the CSR significantly as each of these calls executes network call. It was intentionally splited by this if condition (`if(isServer)`)